File: Haskell/euler_003.hs
{- Largest Prime Factor
 -
 - The prime factors of 13195 are 5, 7, 13 and 29.
 -
 - What is the largest prime factor of the number 600851475143 ?
 -
 -
 - Generate a list of primes by filtering the list of odd numbers by those that
 - have no factors other than itself.
 - To factorise, if X squared is greater than N, then N is the only factor.
 - Otherwise if n mod x is 0, x is a factor.
 - Otherwise x is not a factor, continue through the list.
 - The largest prime factor is the last entry in the list.
-}

primes = 2 : 3 : 5 : filter (null . tail . primeFactors) [7,9..]

primeFactors n = factor n primes
 where
    factor n (x:xs)
      | x*x > n         = [n]
      | n `mod` x == 0  = x : factor (n `div` x) (x:xs)
      | otherwise       = factor n xs

largestPrimeFactor n = last $ primeFactors n

File: Lua/euler_003.lua
--[[
--      The Problem: Largest prime factor
--  The prime factors of 13195 are 5, 7, 13 and 29.
--  
--  What is the largest prime factor of the number 600851475143 ?
--      
--      The Solution
--  Generate all primes until Sqrt N and then get the prime factors of N and
--  return the largest one.
--
--  NOTE: This could do with a lot of improvements.
]]

local n = 600851475143
local LPF = 0

local sieve = {}
function buildSieve (n)
   print("Building sieve...")
   print("Generating all possible numbers to n")
   local limit = math.floor(math.sqrt(n))
   for i = 2, limit do
      local prime = true
      for j, v in ipairs(sieve) do
	 prime = prime and not (i % j == 0)
      end
      if prime then
	 sieve[i] = true
      end
   end
   print("Sieve complete")
end

buildSieve(n)
local currPrime = 2
while (n > 1) do
   if (sieve[currPrime]) then
      n2 = n / currPrime
      if (math.floor(n2) == n2) then
	 LPF = currPrime
	 n = n2
      else
	 currPrime = currPrime + 1
      end
   else
      currPrime = currPrime + 1
   end
end

print(LPF)
